src/Agent.Sdk/SecretMasking/RegexSecret.cs

```diff
@@ -21,31 +33,128 @@ public override Boolean Equals(Object obj)
         {
             return false;
         }
-        return String.Equals(m_pattern, item.m_pattern, StringComparison.Ordinal);
+
+        if (!String.Equals(Pattern, item.Pattern, StringComparison.Ordinal) ||
+            !String.Equals(Moniker, item.Moniker, StringComparison.Ordinal) ||
+            m_regexOptions != item.m_regexOptions)
+        {
+            return false;
+        }
+
+        if (m_sniffLiterals == null && item.m_sniffLiterals == null)
+        {
+            return true;
+        }
+
+        m_sniffLiterals.Equals(item.m_sniffLiterals);
+
+        if (m_sniffLiterals.Count !=  item.m_sniffLiterals.Count)
+        {
+            return false; 
+        }
+
+        foreach (string sniffLiteral in m_sniffLiterals) 
+        { 
+            if (!item.m_sniffLiterals.Contains(sniffLiteral))
+            {
+                return false;
+            }
+        }
+
+        return true;
     }
 
-    public override int GetHashCode() => m_pattern.GetHashCode();
+    public override int GetHashCode()
+    {
+        int result = 17;
+        unchecked
+        {
+            result = (result * 31) + Pattern.GetHashCode();
+
+            if (Moniker != null)
+            {
+                result = (result * 31) + Moniker.GetHashCode();
+            }
+
+            result = (result * 31) + m_regexOptions.GetHashCode();
+
+            // Use xor for set values to be order-independent.
+            if (m_sniffLiterals != null)
+            {
+                int xor_0 = 0;
+                foreach (var sniffLiteral in m_sniffLiterals)
+                {
+                    xor_0 ^= sniffLiteral.GetHashCode();
+                }
+                result = (result * 31) + xor_0;
+            }
+        }
 
-    public IEnumerable<ReplacementPosition> GetPositions(String input)
+        return result;
+    }
+
+    public IEnumerable<Replacement> GetReplacements(String input)
     {
-        Int32 startIndex = 0;
-        while (startIndex < input.Length)
+        bool runRegexes = m_sniffLiterals == null ? true : false;
+
+        if (m_sniffLiterals != null)
         {
-            var match = m_regex.Match(input, startIndex);
-            if (match.Success)
+            foreach (string sniffLiteral in m_sniffLiterals)
             {
-                startIndex = match.Index + 1;
-                yield return new ReplacementPosition(match.Index, match.Length);
+                if (input.IndexOf(sniffLiteral, StringComparison.Ordinal) != -1)
```

Comment:
>IndexOf [](http://example.com/codeflow?start=26&length=7)

A simple IndexOf check is 15x+ faster than something like RE2 in a typical perf test. And so using it as a first pass filter to decide whether to run a follow-on regex is likely to yield performance gains. These are typically 4o% or more in other scan scenarios 1ES owns that use the technique.